pytorch_lightning/loops/fit_loop.py

```diff
@@ -0,0 +1,285 @@
+            if met_min_epochs and met_min_steps:
+                should_stop = True
+            else:
+                log.info(
+                    'Trainer was signaled to stop but required minimum epochs'
+                    f' ({self.min_epochs}) or minimum steps ({self.min_steps}) has'
+                    ' not been met. Training will continue...'
+                )
+        self.trainer.should_stop = should_stop
+
+        return stop_steps or should_stop or stop_epochs
+
+    @property
+    def skip(self) -> bool:
+        """Whether we should skip the training and immediately return from the call to :meth:`run`."""
+        return self.done or self.trainer.num_training_batches == 0
+
+    def connect(self, trainer: 'pl.Trainer', *args: Any, **kwargs: Any) -> None:
+        """Connects the loop with necessary arguments like the trainer"""
+        # TODO(@justusschock): Do we want to forward *args and **kwargs to the inner loop here?
+        # TODO(@justusschock): Can we make the trainer a weakref/proxy?
+        void(*args, **kwargs)
+        self.trainer = trainer
+        self.training_loop.connect(trainer)
+
+    def reset(self) -> None:
+        """Resets the internal state of this loop"""
+
+    def on_run_start(self) -> None:
+        """Calls the ``on_train_start`` hook."""
+        self.results.to(device=self.trainer.lightning_module.device)
+        self.trainer.call_hook("on_train_start")
+
+    def on_advance_start(self) -> None:
+        """Prepares the dataloader for training and calls the hooks ``on_epoch_start`` and ``on_train_epoch_start``"""
+        model = self.trainer.lightning_module
+
+        # reset train dataloader
+        if self.current_epoch != 0 and self.trainer.reload_dataloaders_every_epoch:
+            self.trainer.reset_train_dataloader(model)
+
+        # TODO: specify the possible exception
+        with suppress(Exception):
+            # set seed for distributed sampler (enables shuffling for each epoch)
+            self.trainer.train_dataloader.sampler.set_epoch(self.current_epoch)
+
+        # changing gradient according accumulation_scheduler
+        self.trainer.accumulation_scheduler.on_train_epoch_start(self.trainer, self.trainer.lightning_module)
+
+        # stores accumulated grad fractions per batch
+        self.training_loop.batch_loop.accumulated_loss = TensorRunningAccum(
+            window_length=self.trainer.accumulate_grad_batches
+        )
+
+    def advance(self) -> None:
+        """Runs one whole epoch."""
+        train_dataloader = self.trainer.accelerator.process_dataloader(self.trainer.train_dataloader)
+        train_dataloader = self.trainer.data_connector.get_profiled_train_dataloader(train_dataloader)
+
+        with self.trainer.profiler.profile("run_training_epoch"):
+            # run train epoch
+            epoch_output = self.training_loop.run(train_dataloader)
+
+            if epoch_output is None:
+                return
```

Comment:
@awaelchli do you remember why this was necessary?

Reply:
there is a good reason, something along those lines of my other comment. pretty sure without this a test will fail. I would not have added it otherwise. But I should have added a comment. De-cyphering handling of these edge cases was definitely a center point of frustration for these loop refactors.

Reply:
pretty sure this: https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html?highlight=skip%20epoch#on-train-batch-start

returning -1 here skipping the epoch